### ISSUE TITLE
Local authentication

### ISSUE THREAD

**Issue by jalberto** [kontena/kontena#1517]
As I understand the only way to add user and authenticate to kontena master is using an Oauth2 provider.

For the case where kontena-master is not exposed to internet and work only in a local network, it must be a local authentication method, Or indicate in the documentation that is required to expose kontena-master to internet.

**Comment by kke** [comment]
It's possible to run your own oauth2 provider inside the grid.

**Comment by jalberto** [comment]
@kke that is my 1st option, yet I think a very simple local auth mechanism will be great. This is not a question, more a feature request.

**Comment by SpComb** [comment]
> For the case where kontena-master is not exposed to internet and work only in a local network

AFAIK oauth2 does not require that the master is exposed to the internet? All of the auth flow passes through the user's CLI and browser, using HTTP redirects. It's enough to have a browser that is able to access both the oauth2 provider on the internet, and the internal master.

**Comment by jalberto** [comment]
@SpComb maybe I understood it wrongly, but master needs to recevie callbacks right?

**Comment by SpComb** [comment]
> @SpComb maybe I understood it wrongly, but master needs to recevie callbacks right?

Yes, but AFAIK those callbacks from the oauth provieder come in the form of HTTP `Location: http://master-uri/...` redirects that the browser follows.

**Comment by jalberto** [comment]
@SpComb I just tried a fresh kontena installatoin in a private network connected through a VPN.

The initial auth using the token works as expected.
The cloud login command also succeed in the CLI (no errors messages) but the browser opens this url: 

http://localhost:38945/cb?code=b09c05c734734624b367....

but my kontena-master is not in loclahost, but in a docker-compose installatino in a remote server with a private IP and a internal name (km.foo.loc) this name is resolvable by my laptopt of course.

So maybe is this a bug?

**Comment by kke** [comment]
The master needs to be able to query for userinfo from the auth provider.

The authentication flow (which has a ridiculous amount of steps) is documented in https://www.kontena.io/docs/using-kontena/authentication

Copying here because there's no anchors in the documentation:

1. An administrator of the Kontena Master has created an invitation for an user and has obtained an invitation code.
2. The user issues the command `kontena master join <https://master_url> <invitation_code>`. If the email address in the invitation matches the one used on the authentication provider, the user can also use the normal Master login: `kontena master login <https://master_url>` without an invitation code.
3. The CLI automatically starts a local web server in the background to listen for the OAuth2 callback on a random TCP port. This server is only accessible through the localhost interface and is not exposed to the internet. The final step of the authentication flow will redirect the user to `http://localhost:<port>/cb?code=<auth_code>` and the CLI's local web server then receives the authorization code from the parameters in that request.
4. The CLI requests `https://master_url/authenticate?redirect_uri=http://localhost:<random_port>/cb&invite_code=<invite_code>`
5. Master validates the parameters and tries to find the user by using the invitation code. If an invitation code was not supplied, then the user matching will be performed in the userinfo step by comparing the email address supplied by the auth provider with one in the local user database.
6. The Master creates an AuthorizationRequest record with a random `state` id. This will be used to match the callback coming from the auth provider with the original authentication request.
7. The Master responds with a redirect to the authentication provider's authorization URL.
8. The CLI receives a response to the request from step 4 and opens a browser to the returned authorization URL.
9. The user sees a consent prompt asking for agreement to give the Master access to the user's basic information.
10. The user clicks Agree and the authentication provider redirects the user back to `https://master_url/cb?code=<authorization_code>&state=<state>`
11. The Master parses the query parameters and finds the existing authorization request using the state parameter.
12. The Master exchanges the authorization code for an access token from the authentication provider's token endpoint.
13. The Master requests user information using the received access token and updates the user's external id and username/email in the local user database.
14. The Master creates a local access token with an authorization code.
15. The Master responds with a redirect to `http://localhost:<random_port>/cb?code=<authorization_code>.
16. The user's browser follows the redirect to the local web server, which then parses the code from the query parameters.
17. The local web server is terminated.
18. The CLI exchanges the authorization code for an access token with the Master.
19. The access token is saved to the client's `$HOME/.kontena_client.json` file.
20. The CLI has now been authenticated to access the Master API.

## tldr

1) Browser starts a web server listening in localhost:random_port
2) Browser contacts master and tells it the port of local web server
3) Browser is redirected to auth provider
4) Authentication / authorization is performed
5) Browser is redirected to master
6) Master asks for user information from auth provider
7) Browser is redirected to the localhost webserver from step 1 which parses the auth info from the query parameters in the redirect.

**Comment by jalberto** [comment]
@SpComb clearly I missed that section in the docs, thanks!

Yet my suggestion stands, I think a simpler auth method only for local will be great.

To next user with same doubts I have a suggestion that can help:

- render a simple message in the localhost callback: "your auth succeed" right now, the browser display a non-found error
